docs: metrics (#71778)
metabase/metabase: docs/data-modeling/metrics.md

---

# Metrics

Metrics define the official way to calculate important numbers for your team.

![Metric home page](./images/metric.png)

Metrics are like pre-defined calculations: create your aggregations once, save them as metrics, and use them whenever you need to analyze your data.

For example, you may want to create a metric that calculates revenue, so people can refer to revenue in their own questions. That way you standardize how revenue is calculated (so you don't end up with five different calculations for the revenue).

## How metrics work

Think of a metric like a saved reusable formula. For example, you can create a metric on a `Survey` table to save the calculation `(CountIf([Score] >= 9) - CountIf([Score] <= 6)) / Count()` (aka Net Promoter Score). Other people in you org will be able to just select the metric in the query builder when they want to compute NPS instead of writing their own calculations.

Metrics can be used in questions built with the [graphical query builder](../questions/query-builder/editor.md).

Metrics have three components:

> - [**Data source**](#metric-data-sources): metrics are attached to a specific data source and can be used in any question which is built on that data source.

Reviewer: ```suggestion
- [**Data source**](#metric-data-sources): metrics are associated with a specific data source and can be used in any question built on that data source.
```

## Metric data sources

Metrics are attached to a specific data source. This means that if you define a "Net Promoter Score" metric on the `Survey` table, you will be able to use the metric in any question that is build on the `Survey` table, but _only_ in questions that are built on the `Survey` table (and not, for example, any model or saved question that itself uses the `Survey` table).

You can build metrics on tables, saved questions, models, and other metrics.

We don't recommend using joins in metric definitions. If you need to build a metric that uses information from multiple tables, consider making a [transform](../data-studio/transforms/transforms-overview.md) joining tables, and then create a matric on the output of that transform.

## Metric formula

The formula is the core of the metric. It's the thing you are aggregating, and it's required when defining a metric.

You can use [custom expressions](../questions/query-builder/expressions.md) to define a metric, for example:

```
 (CountIf([Score] >= 9) - CountIf([Score] <= 6) ) / Count() * 100
```

> Metric formula has to be an _aggregation_ , meaning you can have a formula like `Sum([Price])` which will sum all the values in the `Price` column across all records, but you can't have a formula like `[Price] + [Tax]` which adds two values on a per-record basis. If you need to create a "calculated" column, consider using [transforms](../data-studio/transforms/transforms-overview.md) or [models](../data-modeling/models.md) instead.

Reviewer: Emphasize the column v row thing, and rephrase "to avoid slavic tendency to drop article"

```suggestion
Metric formulas must be an _aggregation_  of a column, meaning you can have a formula like `Sum([Price])` which will add up all the values in the `Price` column across all records, but you can't have a formula like `[Price] + [Tax]` which adds two values on a per-record basis. If you need to create a "calculated" column, consider using [transforms](../data-studio/transforms/transforms-overview.md) or [models](../data-modeling/models.md) instead.
```

To **view** a computed metric, people need to have:

- "View data"" permissions that aren't "blocked" on the metric's data source;
- "Curate" permissions on the collection where the metric is saved.

To **use** a metric in the query builder, people need to have:

- "View data" permission on the data source that isn't "blocked";
- At least "Query builder" Create queries permissions;
- "Curate" permissions on the collection where the metric is saved.

Anyone who has these permissions will be able to interact with the metric: for example, anyone with curate permissions to a collection, and "Create queries" permission on a data source will be able to create metrics in the collection that they can curate.

## Metrics in the Library

By adding metrics to your Library, the query builder will display the metrics more prominently in menus. See [Library metrics](../data-studio/library.md#metrics).

## Verifying a metric

On [Pro](https://www.metabase.com/product/pro) and [Enterprise](https://www.metabase.com/product/enterprise) plans, you can [mark metrics as verified](../exploration-and-organization/content-verification.md).

> Prefer using [Library metrics](../data-studio/library.md#metrics) instead of content verification.

Reviewer: ```suggestion
Prefer using the [Library](../data-studio/library.md#metrics) to let people know which metrics they should use.
```

## Limitations

- For now, metrics are attached to a specific data source, and can't be used on any other data source, including on data sources derived from metric's original data source.
- Metrics can only be used in the graphical query builder. See [SQL Snippets](../questions/native-editor/snippets.md) for defining reusable bits of SQL queries.

> - You can't limit people from creating metrics on data sources they otherwise have permissions to query.

Reviewer: Good note; seems like a control we could add to the product.
